code/__HELPERS/icons.dm

```diff
@@ -1055,8 +1055,57 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
 
 GLOBAL_LIST_INIT(freon_color_matrix, list("#2E5E69", "#60A2A8", "#A1AFB1", rgb(0,0,0)))
 
-/obj/proc/make_frozen_visual()
-	// Used to make the frozen item visuals for Freon.
+/atom/proc/begin_freeze()
+	SIGNAL_HANDLER
+	addtimer(CALLBACK(src, .proc/make_frozen), 3 SECONDS)
+
+/atom/proc/begin_unfreeze()
+	SIGNAL_HANDLER
+	if(HAS_TRAIT(src, TRAIT_ON_ICE)) //something is freezing us, don't unfreeze
+		return
+	addtimer(CALLBACK(src, .proc/make_unfrozen), 3 SECONDS)
+
+/atom/proc/make_frozen()
+	return
+
+/atom/proc/make_unfrozen()
+	return
+
+/mob/living/carbon/make_frozen()
+	if(!HAS_TRAIT(src, TRAIT_ON_ICE)) //nothing is freezing us, don't freeze
+		return
+	for(var/obj/item/organ/O in internal_organs)
+		ADD_TRAIT(O, TRAIT_ON_ICE, "owner body temp")
+		O.make_frozen() //no delay
+
+/mob/living/carbon/make_unfrozen()
+	if(HAS_TRAIT(src, TRAIT_ON_ICE)) //something is freezing us, don't unfreeze
+		return
+	for(var/obj/item/organ/O in internal_organs)
+		REMOVE_TRAIT(O, TRAIT_ON_ICE, "owner body temp")
+		O.make_unfrozen() //no delay
+
+/mob/living/carbon/human/make_frozen()
+	if(!HAS_TRAIT(src, TRAIT_ON_ICE))
+		return
+	if(HAS_TRAIT_FROM(src, TRAIT_COLD_SKIN, "make_frozen()")) //already have it, don't need to apply it again
```

Comment:
dont name trait sources after functions like this its hard to work with. even if that source is unique to that proc it should be something like "frozen solid" or whatever